[Registry] Combine with existing registry drafts

citelao opened the pull request:
We have an existing working draft of registry helpers that we've been iterating for a while, so trying to integrate that here.

ChrisGuzak committed:
add wil::str_raw_ptr for winrt::hstring (#277)

citelao committed:
pull change

citelao committed:
various fixes and adjustments, still not building

citelao committed:
remove unused

citelao committed:
clang fixes?

citelao committed:
get it building

citelao committed:
final build errors

citelao committed:
start using these in tests

citelao committed:
potentially update dword calls to support subkey

citelao committed:
remove default_value that wasn't used

citelao committed:
fix all? annotations

citelao committed:
update and fix tests

citelao committed:
fix some sal annotations

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+#ifndef __WIL_REGISTRY_BASIC_INCLUDED
+#define __WIL_REGISTRY_BASIC_INCLUDED
+
+// TODO: should we use feature checks instead of forced imports?
```
@keith-horton What are your thoughts on importing stuff here versus using feature-test macros?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+                return RRF_RT_REG_MULTI_SZ;
+            case REG_BINARY:
+                return RRF_RT_REG_BINARY;
+            // the caller can directly specify their own flags if they need to
```
I have mixed feelings about this function. I think the purpose is that we can specify flags for getters/setters in a simpler way, but should we just be explicit about flags at each call site? That might not be super simple, but worth considering.

Certainly I think we should get rid of the `default` behavior---I think that's sneaky behavior and will fail silently if callers pass in the wrong `type`. Thoughts?

citelao replied:
(Also see `get_access_flags` below, which fails fast instead)

keith-horton replied:
I built this in as most users are familiar with the basic types (REG_DWORD), not the new types (RRF_RT_REG_DWORD) -- or the new semantics in how they interact with the native APIs. I prefer this as it doesn't demand users know new types - they can continue to use the same types that are in their existing code.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+    }
+
+    template <typename err_policy>
+    ::wil::unique_hlocal_ptr<SECURITY_DESCRIPTOR> create_security_descriptor(_In_opt_ PCWSTR security_descriptor)
```
Aside: we'll have to do a pass on this PR to make sure we prefix everything. I wish that were a linter test...

keith-horton replied:
Yeah. I made updates just now for std and wil namespace references.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+template <typename T>
+struct optional_value
+{
+    constexpr bool empty() const WI_NOEXCEPT
```
Should become `has_value` to match `std::optional`.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+};
+
+template <typename T>
+struct optional_value
```
I was discussing this with a coworker---we use this class instead of `std::optional` because (1) we want to be exception-safe and (2) we want to store large values on the heap. Is that right?

Is the expected API:

* `bool has_value()`
* `operator->`, `operator*`
* `T value()`
* `T value_or()` (existing)
* (maybe some of the other weird functions they have?)
https://en.cppreference.com/w/cpp/utility/optional

citelao replied:
Also:

* `operator==`, esp when comparing to raw `T` and `std::nullopt`

citelao replied:
Should we also have a flag that simply consumes `std::optional` here?

keith-horton replied:
The original reason was that I didn't want to *require* all users to include STL headers - as that suddenly cuts down on the total number of users who could make use of this. I'm open to checking if std::optional is available, and if so, just using that type as the name 'optional_value'.

citelao replied:
I wonder if in that case we can write `::wistd::optional`?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+        {
+            // reset to size 0 in case resize() throws
+            string.value.clear();
+            string.value.resize(byteSize / 2);
```
Should this be `sizeof(wchar_t)`?

keith-horton replied:
sure. it can.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+            string.value.resize(byteSize / 2);
+        }
+        catch (...)
+        {
```
Should return false if the resize failed?

citelao replied:
Ditto for other throwing fns.

citelao replied:
We should also handle errors in other non-throwing versions.

citelao replied:
Oh, I'm not sure this is applicable---we test for size `0` after our resizes to validate success. Feel free to resolve.

keith-horton replied:
grow_buffer returns true if growing that buffer type is supported, false if growing that type is not supported. I can rename this to make more sense - maybe grow_buffer_if_supported.

keith-horton replied:
Correct, the check if the alloc succeeded is done by the caller of grow_buffer by checking the new buffer_size

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+
+    // type T is the owning type for storing a key: HKEY, wil::unique_hkey, wil::shared_hkey
+    template <typename T, typename err_policy = err_exception_policy>
+    class reg_view_t
```
I was thinking about this a bit. Is there a downside to simply adding these methods to `unique_hkey`?

There's not a _ton_ of precedent for this, but:

1. It's free. Callers who don't use the functions don't really pay anything simply because the functions exist.
2. It simplifies registry access dramatically. How do I get stuff from a `unique_hkey`? Call methods on it.
3. It matches other languages (like .NET's [`Microsoft.Win32.RegistryKey`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.win32.registrykey.getvalue?view=net-7.0))

citelao replied:
I'm not sure if that aligns with the general feedback you got on this (make the view owning(?))

keith-horton replied:
Right. One option was just change unique_key completely to refer to this object instead - thus won't break any existing users, only adding value (assuming I build something correctly to keep the overall object small, no v-tables, etc).

Raymond had a slightly different suggestion, but I'm still not really sure exactly what that looks like.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+            return err_policy::HResult(HRESULT_FROM_WIN32(error));
+        }
+
+        // TODO: just call set_value with nullptr subkey?
```
@keith-horton I added the variants that accept a `subkey`. Normally I adjust the non-subkey version to call the subkey version with `nullptr`, but these setters are using `RegSetValueExW` instead of `RegSetKeyValueW` (and that doesn't take a subkey param).

The [documentation for `RegSetValueExW`](https://learn.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regsetvalueexw#remarks) says: 

> Consider using the [RegSetKeyValue](https://learn.microsoft.com/en-us/windows/desktop/api/winreg/nf-winreg-regsetkeyvaluew) function, which provides a more convenient way to set the value of a registry key.

So should we just always use `RegSetKeyValueW`?

keith-horton replied:
I don't know I have a strong opinion on which of the 2 APIs to use. There are only 2 other callsites, so we can just use RegSetKeyValueW

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+    return set_value_string_nothrow(key, nullptr, value_name, data);
+}
+
+// TODO: write multisz setters that don't throw.
```
I had to disable these because they caused errors when compiling with clang. I believe they call functions which use `try/catch`, which isn't supported in exceptions-off scenarios.

keith-horton replied:
crap. yeah. I bet there are various sites which I didn't ifdef correctly for Exceptions or for STL support

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+#ifdef WIL_ENABLE_EXCEPTIONS
+inline void set_value_dword(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, DWORD data)
+{
+    const reg_view_details::reg_view regview{key};
```
We discussed this, but trying to capture it here: 

We should see if we can invert this calling structure---IMO, the `reg_view` should call the free functions to get the data. That way, folks calling the free functions directly do not need to pay the cost of instantiating a reg_view.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+        }
+
+        template <typename R>
+        typename err_policy::result get_value(_In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, optional_value<R>& return_value, DWORD type = ::wil::registry::reg_view_details::get_value_type<R>()) const
```
We discussed this, but I want to separate this into `get_value` and `try_get_value`, where `get_value` errors if the value is missing.

That way callers that expect a value to exist can fail quicker.

keith-horton replied:
we should probably discuss this more - it looks like wil tend to return a bool (or something convertable to bool) from its try_* functions. If that's the case, we'll have to rework these.

Looking through this again, I think I want to make this more consistent so that get* calls either return:
- HRESULT for no-throw, and they pass an OUT ptr for the type they want to read (if they really want to deal with all errors, including value-not-found).
- bool for try_get (or maybe return optional? since that can be convertable to bool?)
- not sure what to return now for just get_value() that doesn't return an HRESULT if try_get_value is returning an optional. I don't know that I want to have the 'default' get_value throw when the value isn't present.

citelao replied:
My hunch is that the prior art in wil was for lack of a good `optional` type. It's pretty idiomatic to return an `optional<T>` in modern C++, plus it avoids the outptr shenanigans.

I think the approach we've ended up with is good:

* `*_nothrow` returns `HRESULT` (including not found), uses outptr.
* `try_get_*` returns `optional<T>` (empty when not found; throws on other errors), no outptr.
* `get_*` returns `T` (throws on all errors, including not found), no outptr.

This makes sense: if a caller expects a value to exist, it is an error for it not to exist.

citelao replied:
(Since we talked about it: the Rusty way would be to always return an optional and have callers `.expect()` it 😏 )

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+    return set_value_qword(key, nullptr, value_name, data);
+}
+
+inline void set_value_string(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, _In_ PCWSTR data)
```
We need to support variants that take `std::wstring` (and potentially other string types/string_view).

Why? Registry strings can have embedded nulls 🤣, so a PCWSTR can lead to data loss.

keith-horton replied:
sadly, std::wstring cannot have embedded nulls either (at least to be conformant). IMHO, shoving something with embedded nulls as a REG_SZ should just go through the raw byte* version - it should stick out badly as "I'm doing something super-strange and not recommended, so I have to use the raw API to do it, not a friendly API" (I don't want to suggest this is a good idea by having friendly support for it).

citelao replied:
I think that's a good point. We shouldn't make it easy to do Bad Things.

However, I know that C++/WinRT is deliberate to recommend string_views in some cases when there's interop with HSTRINGs, in part because HSTRINGs can have embedded nulls, right?

I recommend at least including a note indicating our behavior for this edge case (truncation, right?). Optimally we'd have some sort of variant call pattern that returns a vector of bytes or something (or an HSTRING or something?)

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+    return regview.set_value(subkey, value_name, data);
+}
+
+inline HRESULT set_value_string_nothrow(HKEY key, _In_opt_ PCWSTR value_name, _In_ PCWSTR data) WI_NOEXCEPT
```
As I mention above for the throwing version, since registry values can have nulls in them, we need to support alternate string types (`hstring`?) even in the nothrow.

keith-horton replied:
yeah. we should talk about how to inform the user "hey, this value is REG_SZ but it turns out someone is doing weird stuff by having embedded nulls and not just using REG_MULTI_SZ, or someone is trying to haxxor you!".

And if they *really* want to read the whole thing and parse out the embedded strings, they can call the raw binary version - since they'll need to be really careful.

or if we think this is actually a common case, we can do the parsing and let them get back a vector<wstring>.

citelao replied:
I'm gonna resolve this conversation so we can keep one cannonical convo about embedded nulls, since I keep asking about it lol

citelao reviewed (commented):
Left some initial feedback as we work

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+//*********************************************************
```
Should this be a separate file, or should it be auto-included in `registry.h`?

keith-horton replied:
I'm tempted right now to put the "supported functions" directly in registry.h, but have them call into registry_basic.h. Until we have everything we want, then we won't need registry_basic to 'hide' the details. (or maybe we will prefer it that way and just keep the extensive template goo in registry_basic.h and keep registry.h more readable / searchable).

in which case registry_basic.h might be better named registry_details.h ?

citelao replied:
We discussed offline, but I recommend simply using one file, `registry.h`. I'm not sure it really simplifies things to extract the functions. Frankly, I'd sooner move the registry watcher.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+{
+namespace registry
+{
+constexpr bool error_value_was_not_found(HRESULT hr) WI_NOEXCEPT
```
These seem to be unused

keith-horton replied:
yeah. they are on the public wil::registry namespace. i.e. friendly functions if callers don't want to hunt down error codes themselves.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1540 @@
+                    if (::wil::registry::reg_view_details::get_buffer_size(return_value) == 0)
+                    {
+                        error = ERROR_NOT_ENOUGH_MEMORY;
+                        break;
```
Should this be an early return, not a break? It looks like this error is ignored.

keith-horton replied:
oops. this is a bug. thanks! fixing.

citelao committed:
initial batch of try helpers

citelao committed:
consume try helpers

citelao committed:
more tests

citelao committed:
leave todo

citelao committed:
another todo

citelao commented on include/wil/registry_basic.h:
```diff
@@ -962,28 +954,86 @@ namespace reg_view_details
             return get_value(nullptr, value_name, return_value, type);
         }
 
         // intended for err_exception_policy as err_returncode_policy will not get an error code
```
Should we use enable_ifs for these functions? And below.

keith-horton replied:
@citelao , would that give better compiler errors than what would be provided already? I'm not really sure :(

citelao commented on include/wil/registry_basic.h:
```diff
@@ -962,28 +954,86 @@ namespace reg_view_details
+            } else if (hr == HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND))
+            {
+                return_value.type = optional_value_status::no_value;
+                // TODO: reset value.
```
Do we need to explicitly clear the `.value` of the `optional`? I think we probably do.

keith-horton replied:
yep!

citelao commented on include/wil/registry_basic.h:
```diff
@@ -962,28 +954,86 @@ namespace reg_view_details
+            if (SUCCEEDED(hr))
+            {
+                return_value.type = optional_value_status::has_value;
+                // TODO: set value
```
I think the value has already been set at this point, so we may not have any other action to take.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1675 @@
+// Registry Set* functions, both throwing and non-throwing
+
+#ifdef WIL_ENABLE_EXCEPTIONS
+inline void set_value_dword(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, DWORD data)
```
An idle thought: why `set_value_dword` and `get_value_dword`---why not `set_dword` and `get_dword`? I know we're setting values, but calling looks like: `wil::registry::set_value_dword` which is a bit of a handful.

keith-horton replied:
that's a good point. we should try it out and see how it reads (probably reads just fine)

citelao commented on include/wil/registry_basic.h:
```diff
@@ -0,0 +1,1675 @@
+
+namespace wil
+{
+namespace registry
```
I see 3 options for "namespacing" this:

1. (this approach) `wil::registry`
2. Abbreviated: `wil::reg`
3. Prefixed: `wil::reg_*`
4. None: `wil::*` (`wil::create_reg_key`, `wil::set_reg_dword` eg)

The current approach works well enough and is super clear, but it leads to code like this:

```cpp
const auto key = wil::registry::open_key(/* ... */);
const auto value = wil::registry::get_value_dword(key, /* ... */);
wil::registry::set_value_dword(key, /* ... */);
// etc
```

vs `wil::reg::get_dword` etc might be easier to type.

Interested on all thoughts here.

keith-horton replied:
I personally prefer typing out 'registry' as a namespace. but I do see your point. It looks like the registry watcher code doesn't use a namespace.

if I had to pick a no-namespace-version, I would probably go with:
wil::open_reg_key()
wil::get_reg_dword()

I think I prefer namespaces, though:
wil::reg::open_key()
wil::reg::get_dword()

I thought putting the COM stuff in its own namespace made things much better

citelao replied:
I like `wil::reg::*` as you've done. I'm happy to stick with that :)

citelao commented on include/wil/registry_basic.h:
```diff
@@ -917,34 +915,35 @@ namespace reg_view_details
                 {
-                    return_value.type = optional_value_status::no_value;
-                    break;
+                    // Simply pass on any not-found errors.
```
"...to the err_policy"

citelao commented on include/wil/registry_basic.h:
```diff
@@ -957,24 +956,102 @@ namespace reg_view_details
+
+        // intended for err_exception_policy as err_returncode_policy will not get an error code
+        template <typename R>
+        R try_get_value(_In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, DWORD type = ::wil::registry::reg_view_details::get_value_type<R>()) const
```
Try should return an optional

keith-horton replied:
Yeah. I think I messed this up. We can talk more - see my comment above :)

keith-horton committed:
Addressing feedback.

UMU618 committed:
Fix doc typo (#284)

keith-horton committed:
Fixed a lot of issues.
All current registry tests now pass.

citelao commented on tests/BasicRegistryTests.cpp:
```diff
@@ -230,4 +230,95 @@ TEST_CASE("BasicRegistryTests::Strings", "[registry][get_registry_string]")
+        //std::wstring value = L"something \0 pithy";
+        std::wstring value = L"something pithy";
+        REQUIRE_SUCCEEDED(wil::reg::set_value_string_nothrow(HKEY_CURRENT_USER, testSubkey, stringValueName, value.c_str()));
+        auto result = wil::reg::get_value_string(HKEY_CURRENT_USER, testSubkey, stringValueName);
```
This isn't testing multistring, right? (And all subsequent tests)

citelao commented on include/wil/registry_basic.h:
```diff
@@ -33,19 +35,19 @@
+namespace reg
 {
-constexpr bool error_value_was_not_found(HRESULT hr) WI_NOEXCEPT
+constexpr bool is_value_was_not_found(HRESULT hr) WI_NOEXCEPT
```
Perhaps `is_error_not_found` and `is_error_more_data` (or `is_response_not_found` and `is_response_too_large`)?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -791,247 +798,226 @@ namespace reg_view_details
-#endif
-
-        ::wil::unique_hkey open_unique_key(_In_opt_ PCWSTR subKey, ::wil::registry::key_access access = ::wil::registry::key_access::read) const
+        ::wil::unique_hkey open_unique_key(_In_opt_ PCWSTR subKey, ::wil::reg::key_access access = ::wil::reg::key_access::read) const
```
Shouldn't these factory functions be static?

citelao replied:
The more I think about this, the more I'm convinced this is definitely true---the current design requires 2-phase initiailization:

```cpp
const reg_view_details::reg_view regview{key}; // invalid here
regview.open_unique_key(path, access); // valid here
```

And [the C++ Core Guidelines discourage that](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#nr5-dont-use-two-phase-initialization)

citelao commented on include/wil/registry_basic.h:
```diff
@@ -791,247 +798,226 @@ namespace reg_view_details
+                    ::wil::reg::reg_view_details::get_buffer(return_value),
                     &data_size_bytes);
-                if (error == ERROR_SUCCESS)
+                printf("\n\tRegGetValueW returned error %d (data_size_bytes %lu, buffer pointer %p)\n", error, data_size_bytes, ::wil::reg::reg_view_details::get_buffer(return_value));
```
Is it useful to convert this into a `sprintf` and include that error message as context if the call fails?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -791,247 +798,226 @@ namespace reg_view_details
-        typename get_err_policy::result get_value_internal(_In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, R& return_value, DWORD type = ::wil::registry::reg_view_details::get_value_type<R>()) const
+        typename get_err_policy::result get_value_internal(_In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, R& return_value, DWORD type = ::wil::reg::reg_view_details::get_value_type<R>()) const
         {
             for (;;)
```
Should we include a way of providing max-iterations for this? This loop can technically spin forever, right?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -791,247 +798,226 @@ namespace reg_view_details
-        // intended for err_exception_policy as err_returncode_policy will not get an error code
         template <typename R>
-        R try_get_value(_In_opt_ PCWSTR value_name, DWORD type = ::wil::registry::reg_view_details::get_value_type<R>()) const
+        ::wil::reg::optional_value<R>& try_get_value(_In_opt_ PCWSTR value_name, DWORD type = ::wil::reg::reg_view_details::get_value_type<R>()) const
```
We shouldn't return a reference here, right?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -791,247 +798,226 @@ namespace reg_view_details
-        // intended for err_exception_policy as err_returncode_policy will not get an error code
-        template <typename R>
-        R get_value(_In_opt_ PCWSTR value_name, DWORD type = ::wil::registry::reg_view_details::get_value_type<R>()) const
+        ::wil::reg::optional_value<R> try_get_value(_In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, DWORD type = ::wil::reg::reg_view_details::get_value_type<R>()) const
```
Can we include a try_get_value that works without exceptions?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -1040,39 +1026,31 @@ namespace reg_view_details
         }
 
-#ifdef WIL_ENABLE_EXCEPTIONS
+#if defined(__WIL_WINREG_STL)
```
Does `__WIL_WINREG_STL` imply `WIL_ENABLE_EXCEPTIONS`, or do we still need to check both in places like this?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -1156,14 +1118,14 @@ namespace reg_view_details
+            return value_enumerator(::wil::reg::details::get_key(m_key));
         }
-#endif
+#endif // #if defined(__WIL_WINREG_STL
```
Nit: missing `)`

citelao commented on include/wil/registry_basic.h:
```diff
@@ -1172,85 +1134,108 @@ namespace reg_view_details
+inline ::wil::unique_hkey try_open_unique_key(HKEY key, _In_opt_ PCWSTR path, ::wil::reg::key_access access = ::wil::reg::key_access::read)
+{
+    const reg_view_details::reg_view_nothrow regview{ key };
+    return regview.open_unique_key(path, access);
```
Does this work? Shouldn't a nothrow call return an error code, not a `::wil::unique_hkey`?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -1378,114 +1365,125 @@ inline HRESULT set_value_string_nothrow(HKEY key, _In_opt_ PCWSTR value_name, _I
+    return set_value_multistring_nothrow(key, nullptr, value_name, data);
+}
+
+inline HRESULT set_value_multistring_nothrow(HKEY key, _In_opt_ PCWSTR subkey, _In_ PCWSTR value_name, const ::std::list<::std::wstring>& data) WI_NOEXCEPT
```
Should we:

* Just support `std::vector` and let callers move to `std::list` themselves?
* Just support `std::list` for calls like this?
* Template this function to support any enumerable type?
* Something else?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -1378,114 +1365,125 @@ inline HRESULT set_value_string_nothrow(HKEY key, _In_opt_ PCWSTR value_name, _I
     const reg_view_details::reg_view regview{key};
-    return regview.get_value<DWORD>(subkey, value_name);
+    DWORD return_value{};
+    regview.get_value<DWORD>(subkey, value_name, return_value);
```
I'd love to still support the `R get_value()` functions that return the value and not an error code, if exceptions are enabled.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -1531,133 +1528,123 @@ inline HRESULT get_value_dword_nothrow(HKEY key, _In_opt_ PCWSTR value_name, _In
+{
+    const reg_view_details::reg_view_nothrow regview{ key };
+    ::wil::reg::optional_value<::std::vector<BYTE>> registry_value {regview.try_get_value<::std::vector<BYTE>>(subkey, value_name, type)};
+    RETURN_IF_FAILED(registry_value.inner_error);
```
I'm not sure I like the `inner_error` concept---I'd much rather use a straightforward `std::optional` as an outparam, then have the function return an HRESULT.

This seems a lot like Rust's `Result` type, which is really cool, but I think it's a bit non-standard for C++. I'd rather stick to common paradigms and use std-like types whenever possible.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -1531,133 +1528,123 @@ inline HRESULT get_value_dword_nothrow(HKEY key, _In_opt_ PCWSTR value_name, _In
 }
 
-inline HRESULT get_value_multistring_wstring_nothrow(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, ::std::vector<::std::wstring>* return_value) WI_NOEXCEPT try
+inline HRESULT get_value_multistring_wstring_nothrow(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, _In_ ::std::vector<::std::wstring>* return_value) WI_NOEXCEPT try
```
These pointer params (including the ones that were formerly references) are all `_Out_`

citelao commented on include/wil/registry_basic.h:
```diff
@@ -1531,133 +1528,123 @@ inline HRESULT get_value_dword_nothrow(HKEY key, _In_opt_ PCWSTR value_name, _In
-        return_value.type = optional_value_status::has_value;
-        std::swap(return_value.value, local_value);
+        ::wil::reg::optional_value<::std::vector<::std::wstring>> return_value;
+        return_value.value_assigned = true;
```
`value_assigned` should automatically be set upon inserting/swapping a value into an `optional_value`; it shouldn't be set externally.

citelao reviewed (commented):
Prelim review

AZero13 committed:
Fix Clang-15 warnings (#264)

* Fix Clang-15 warnings

Fix warnings exposed from Clang-Tidy

* Update include/wil/winrt.h

Co-authored-by: Duncan Horn <40036384+dunhor@users.noreply.github.com>

dunhor committed:
Add Tests with Address & UB Sanitizers Enabled (#285)

* Getting address sanitizer to work with MSVC and compile with Clang

* Some CMake fixups

* Get it working for Clang

* Enable UBSan

* Still need to debug

* Temporarily revert use of /Ob2 due to bug in Clang

* Update tests/sanitize-address/CMakeLists.txt

* Update scripts/azure-pipelines.yml

dmachaj committed:
Port fixes from private repo (#287)

Co-authored-by: Jon Wiswall <jonwis@microsoft.com>
Co-authored-by: Duncan Horn <40036384+dunhor@users.noreply.github.com>

keith-horton committed:
More tests, a few more fixes.

keith-horton committed:
Mirroring the DWORD test cases to QWORD and wstring tests.

oldnewthing committed:
hide_name to intentionally hide a lambda capture or parameter (#198)

One of the sharp edges of C++ coroutines is lifetime management.

A lambda coroutine with captures cannot safely access the captures
after the coroutine suspends, because the lambda may have been
destructed when the coroutine resumes.

A coroutine function cannot safely access any reference parameters
after the coroutine suspends, because the caller may have destroyed
the referenced objects in the meantime.

To avoid around the problem, you have to copy the unsafe things
into your frame before your first suspension.

```cpp
void Widget::RegisterHandlers(int id)
{
    m_button.Click([id, weak_this = get_weak()](auto&& sender, auto&& args)
    {
        // Copy things into the frame for safekeeping before
        // our first co_await.
        if (auto strongThis = weak_this.get())
        {
          auto idCopied = id;

          // After this co_await, "id" and "weak_this" are no longer
          // safe to use.
          co_await winrt::resume_background();

          // Use the captured variables instead.
          strongThis->Update(idCopied);
        }
    });
}

winrt::IAsyncAction Widget::DoSomethingAsync(Thing const& thingRef)
{
    // Copy things into the frame for safekeeping before our first co_await.
    auto strongThis = get_strong();
    auto thing = thingRef;

    // After this co_await, "thingRef" is no longer safe to use.
    co_await winrt::resume_background();

    // Use the captured variables instead.
    strongThing->Method(thing);
}
```

To prevent accidental use of these unsafe variables, you can
intentionally hide them.

```cpp
void Widget::RegisterHandlers(int id)
{
    m_button.Click([id, weak_this = get_weak()](auto&& sender, auto&& args)
    {
        // Copy things into the frame for safekeeping before
        // our first co_await.
        if (auto strongThis = weak_this.get())
        {
          auto idCopied = id;

          // Hide these names so you can't use them by mistake.
          // They are about to become unsafe.
          wil::hide_name id, weak_this; // <<<<<<<<<< NEW

          // After this co_await, "id" and "weak_this" are no longer
          // safe to use.
          co_await winrt::resume_background();

          // Use the captured variables instead.
          strongThis->Update(idCopied);
        }
    });
}

winrt::IAsyncAction Widget::DoSomethingAsync(Thing const& thingRef)
{
    // Copy things into the frame for safekeeping before our first co_await.
    auto strongThis = get_strong();
    auto thing = thingRef;

      // Hide this name so you can't use it by mistake.
      // it is about to become unsafe.
      wil::hide_name thingRef; // <<<<<<<<<<<<<<<<<< NEW

    // After this co_await, "thingRef" is no longer safe to use.
    co_await winrt::resume_background();

    // Use the captured variables instead.
    strongThing->Method(thing);
}
```

C++ permits you to hide names from outer scopes, so a function body
can hide parameters, member variables, functions and methods,
and types defined at outer scopes. A lambda function body can
also hide captured variables.

Implementation details
======================

We define `hide_name` as a function alias, so when you hide a name,
what you're doing is declaring a function with the same name.
Using a function has these benefits:

* Unlike local variables, you don't get "declared but not used"
  warnings for functions on any major compiler (msvc, gcc, clang, icc).
* If you manage to take the address of the function, thinking that
  you're taking the address of a captured variable or parameter,
  you will get an error because function pointers are not implicitly
  convertible to `void*`.
* You can't assign to a function, so attempted assignments are blocked.
* If you try to use the name, it will decay to a function pointer
  that is unusable because it requires you to pass a parameter which
  is an incomplete type.
* If you manage to bypass all these compile-time checks, you get a linker error
   saying that the function `void name(wil::hidden_name)` does not exist.

This is either a clever hack or an abuse of the language.

If you try to read the value of a hidden name:

* MSVC: error C2664: cannot convert from `wil::hide_name (__cdecl *)` to `T`
* MSVC: error C2296: binary operation illegal, left operand has type `wil::hide_name (__cdecl *)`
*	gcc: error: invalid conversion from `void (*)(wil::hidden_name)` to `T` [-fpermissive]
* gcc: warning: pointer to a function used in arithmetic (followed by another error when the result is used)
* gcc: error: invalid operands of types `wil::hide_name` {aka `void(wil::hidden_name)`} and `T` to binary operator
*	clang: error: cannot initialize a variable of type `T` with an lvalue of type `wil::hide_name` (aka `void (struct hidden_name)`)
* clang: error: arithmetic on a pointer to the function type `wil::hide_name` (aka `void (struct hidden_name)`)
* clang: error: invalid operands to binary expression (`wil::hide_name` (aka `void (struct hidden_name)`) and `T`)
* icc: error: value of type "`wil::hide_name *`" cannot be used to initialize an entity of type "`T`"
* icc: error: argument of type "`wil::hide_name *`" is incompatible with parameter of type "`T`"
* icc: error: expression must have integral or unscoped enum type

Assigning to a hidden variable:

*	msvc: error C2659: '=': function as left operand
*	gcc: error: assignment of function 'void v(wil::hidden_name)'
*	clang: error: non-object type `wil::hide_name` (aka `void (struct hidden_name)`) is not assignable
* icc: error: expression `name` must be a modifiable lvalue

Co-authored-by: Jon Wiswall <jonwis@microsoft.com>
Co-authored-by: Duncan Horn <40036384+dunhor@users.noreply.github.com>

keith-horton committed:
Normalized tests - multi-string now passing all tests

keith-horton committed:
Fixing STL header inclusion in registry_basic

dunhor committed:
Sync with OS (#296)

* Sync with OS commit b58d62a22e8b065355e83f17d32f446bac7d5e75

* Fix build issues

* Fix test

citelao commented on tests/BasicRegistryTests.cpp:
```diff
@@ -25,209 +30,1186 @@ TEST_CASE("BasicRegistryTests::Dwords", "[registry][get_registry_dword]")
+    }
+
+    /*
+    SECTION("get and set with string key and value name, template versions")
```
@keith-horton Is this commented out because you were hitting issues getting the templating right?

citelao commented on tests/BasicRegistryTests.cpp:
```diff
@@ -25,209 +30,1186 @@ TEST_CASE("BasicRegistryTests::Dwords", "[registry][get_registry_dword]")
     }
 
-    SECTION("get and set with opened key and value name, nothrow")
+    SECTION("set_value_dword_nothrow/get_value_dword_nothrow: with opened key")
```
Given the repetition here, maybe we should extract template test helpers? I know it's kinda silly, but it might be nice to easily scaffold tests for all types...

citelao replied:
On the other hand, verbosity makes debugging tests easiser.

keith-horton replied:
yes. I copied pasted a ton because debugging tests when there's a lot of shared code is a pain

citelao commented on tests/BasicRegistryTests.cpp:
```diff
@@ -25,209 +30,1186 @@ TEST_CASE("BasicRegistryTests::Dwords", "[registry][get_registry_dword]")
 #endif
 }
 
 #ifdef WIL_ENABLE_EXCEPTIONS
```
Are wstrings exception-safe? I assume some exception-safe code will want to use strings if possible.

citelao replied:
How much work would it be to get strings working exception-free?

keith-horton replied:
std::strings require the caller to have allowed for C++ exceptions. There's not much if any on a basic_string that's noexcept to be safely used without C++ exceptions.

citelao commented on tests/BasicRegistryTests.cpp:
```diff
@@ -25,209 +30,1186 @@ TEST_CASE("BasicRegistryTests::Dwords", "[registry][get_registry_dword]")
+    SECTION("set_value/try_get_value_qword: with open key")
+    {
+        wil::unique_hkey hkey;
+        REQUIRE_SUCCEEDED(wil::reg::create_key_nothrow(HKEY_CURRENT_USER, testSubkey, hkey, wil::reg::key_access::readwrite));
```
We can throw in these execption-enabled tests, right?

citelao commented on tests/BasicRegistryTests.cpp:
```diff
@@ -25,209 +30,1186 @@ TEST_CASE("BasicRegistryTests::Dwords", "[registry][get_registry_dword]")
+        REQUIRE(hr == HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND));
+
+        // fail if get* requests the wrong type
+        DWORD dwordResult{};
```
I'd be interested in explicitly testing reading a *string* from this as the wrong type, too.

keith-horton replied:
this web UI is terrible - I can't really see which test case this is. happy to add whatever test cases you want (or feel free to add them yourself :))

citelao commented on include/wil/registry_basic.h:
```diff
@@ -33,38 +33,38 @@
 {
-    return hr == HRESULT_FROM_WIN32(ERROR_MORE_DATA);
-}
+    constexpr bool is_value_was_not_found(HRESULT hr) WI_NOEXCEPT
```
Nit: do these functions belong in `details`?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
             case REG_BINARY:
                 return RRF_RT_REG_BINARY;
-            // the caller can directly specify their own flags if they need to
+                // the caller can directly specify their own flags if they need to
```
I'm still not a huge fan of this part of the function---I think it gives too much power to callers. But I don't feel super strongly.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
-            if (!tempString.empty())
+            ::std::vector<::std::wstring> strings;
+
+            if (data.size() < 3)
```
`3` is definitely worth a comment :)

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
+
+            auto current = data.cbegin();
+            const auto end_iterator = data.cend();
+            const auto lastNull = (end_iterator - 1);
```
This assumes the data is well-formed, right? Do we need to be careful/defensive about this data?

citelao replied:
Or should we comment that

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
+            while (current != end_iterator)
             {
-                break;
+                const auto next = ::std::find(current, data.cend(), L'\0');
```
multistring strings cannot include embedded nulls, correct?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
-        // increment by integer
-        iterator_t& operator+=(size_t offset)
+        template <>
+        constexpr bool should_return_not_found<::wil::err_exception_policy>()
```
Worth noting that error-code folks can use `try_get_*` variants if they don't want an error there?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
+#if defined(_VECTOR_) && defined(WIL_ENABLE_EXCEPTIONS)
+        // These iterator classes facilitates STL iterator semantics to walk a set of subkeys and values
+        // within an instantiated Key object. This can be directly constructed passing a unique_hkey
+        // but will likely be more commonly constructed from reg_view_t::reg_enum_keys() and reg_view_t::reg_enum_values()
```
I think you're aware, but heads up that we're missing tests for the `enum_*` functions.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
 
-    inline void* get_buffer(const ::std::wstring& string) WI_NOEXCEPT
+    template <typename T>
+    struct optional_value
```
2 nits:

* Does this belong in `wistd`?
* `wistd::optional` instead of `optional_value`, since it's the same concept?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
-    {
-        return const_cast<unsigned long*>(&value);
-    }
+        // TODO: do we need a default version, or should we fail for unknown types?
```
I'd fail, TBH

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
         {
-            auto error = RegDeleteTreeW(::wil::registry::details::get_key(m_key), sub_key);
-            if (error == ERROR_FILE_NOT_FOUND)
+            const auto offset = buffer.find_first_of(L'\0');
```
Can't registry strings include embedded nulls?

citelao replied:
Resolving this; we have another convo for it.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
             }
 
-            [[nodiscard]] ::wil::registry::details::key_iterator end() const WI_NOEXCEPT
+            typename err_policy::result open_key(_In_opt_ PCWSTR subKey, ::wil::unique_hkey& hkey, ::wil::reg::key_access access = ::wil::reg::key_access::read) const
```
Does this class kinda want `static` open and create functions? I know we basically have them, but perhaps it's nice to be able to open a regkey end-to-end with a static method.

`wil::reg::reg_view_nothrow::create()` eg

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
-    return try_get_value_qword(key, nullptr, value_name);
-}
+        // reg_view with the raw HKEY type is non-owning
+        using reg_view_nothrow = reg_view_details::reg_view_t<HKEY, ::wil::err_returncode_policy>;
```
I propose:

* `unique_reg` & `unique_reg_nothrow` (or `unique_reg_view`, etc)
* `shared_reg` & `shared_reg_nothrow`
* `reg_view` & `reg_view_nothrow`

This aligns with the style for unique & shared ptrs, plus it also separtes reg _views_ which are non-owning and reg "things" which are owning.

It also may still be worth re-examining `unique_hkey` et al and adding all these functions there.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
+            typename err_policy::result open_key(_In_opt_ PCWSTR subKey, ::wil::unique_hkey& hkey, ::wil::reg::key_access access = ::wil::reg::key_access::read) const
             {
-                return ::wil::registry::details::key_iterator(m_hkey, ::wil::registry::details::iterator_creation_flag::end);
+                auto error = ::RegOpenKeyExW(::wil::reg::details::get_key(m_key), subKey, 0, ::wil::reg::details::get_access_flags(access), ::wil::out_param(hkey));
```
Do we have a need to support custom `ulOptions`?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
+                auto error = ::RegOpenKeyExW(::wil::reg::details::get_key(m_key), subKey, 0, ::wil::reg::details::get_access_flags(access), ::wil::out_param(hkey));
+                if (error == ERROR_FILE_NOT_FOUND && !::wil::reg::details::should_return_not_found<err_policy>())
+                {
+                    error = ERROR_SUCCESS;
```
Shouldn't this throw in err or failfast variants?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
-            return value_enumerator(::wil::registry::details::get_key(m_key));
-        }
-#endif
+                DWORD disposition = 0;
```
Can we let callers have access to disposition somehow? That's really useful sometimes

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
-            HKEY m_hkey;
-        };
+#if defined(__SDDL_H__)
+            typename err_policy::result create_key(_In_ PCWSTR subKey, ::wil::unique_hkey& hkey, ::wil::reg::key_access access = ::wil::reg::key_access::read, _In_opt_ PCWSTR security_descriptor = nullptr) const
```
Is there a good way to test these SDDL variants?

keith-horton replied:
yeah. I still need to add the create_key tests :(

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
-    const reg_view_details::reg_view_nothrow regview{key};
-    return regview.create_key(path, hkey, access, security_descriptor);
-}
+            typename err_policy::result delete_key(_In_opt_ PCWSTR sub_key) WI_NOEXCEPT
```
Can we use these delete calls in our tests?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
-        nullptr, nullptr, nullptr, nullptr, nullptr, nullptr));
-    return S_OK;
-}
+                    if (error == ERROR_SUCCESS)
```
When does this hit?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
-}
+#if defined(_VECTOR_) && defined(_STRING_) && defined(WIL_ENABLE_EXCEPTIONS)
+            template <typename R>
+            typename err_policy::result set_value_multistring(_In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, const R& data) const try
```
Does this suggest that `set_value` calls won't work for multistring? Is there anyway we could provide this code as a template specialization of `set_value`?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -77,1599 +77,1826 @@ namespace details
+#endif // #if defined(WIL_ENABLE_EXCEPTIONS)
+
+#if defined(_VECTOR_) && defined(WIL_ENABLE_EXCEPTIONS)
+    inline ::std::vector<BYTE> get_value_byte_vector(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, DWORD type)
```
Nit: `get_value_bytes`?

citelao replied:
The vector part is an implementation detail.

citelao committed:
remove besto's original version

keith-horton committed:
using the same template functions for all the set_value* variants.

keith-horton committed:
Updated get_value to use a template version

keith-horton committed:
Finished support for expanded strings + unit tests.

Fixed unit tests for reading fixed-size (array) for reading strings.

keith-horton committed:
Merge branch 'user/besto/registry_helpers_horton_v' of https://github.com/citelao/wil into user/besto/registry_helpers_horton_v

keith-horton committed:
Naming null parameters to Win32 apis.
fixing calls that pass SDDL strings.
adding more tests.

keith-horton committed:
Fixes and unit tests for open key

keith-horton committed:
Cleaning up some create_key calls

citelao commented on runtests.cmd:
```diff
@@ -0,0 +1,4 @@
+.\out\build\x64-Debug\tests\cpplatest\witest.cpplatest.exe BasicRegistryTests*
```
Accidental checkin?

keith-horton replied:
ugh. yes. thanks!

citelao commented on tests/BasicRegistryTests.cpp:
```diff
@@ -12,17 +16,231 @@ constexpr auto* dwordValueName = L"MyDwordValue";
+        REQUIRE(result == 3);
+
+        // fail open if the key doesn't exist
+        hr = wil::reg::open_key_nothrow(hkey.get(), (std::wstring(subSubKey) + L"_not_valid").c_str(), &opened_key, wil::reg::key_access::read);
```
Can you test that `opened_key` is invalid after this call?

citelao commented on tests/BasicRegistryTests.cpp:
```diff
@@ -917,7 +1215,371 @@ TEST_CASE("BasicRegistryTests::wstrings", "[registry][get_registry_string]")
+        // fail if get* requests the wrong type
+        try
+        {
+            wil::reg::try_get_value_dword(hkey.get(), stringValueName);
```
Should this be a nothrow getter?

keith-horton replied:
yes. but now I look at this, I need to rewrite all of these so we're testing the function each test is intended to test :(

citelao commented on tests/BasicRegistryTests.cpp:
```diff
@@ -917,7 +1215,371 @@ TEST_CASE("BasicRegistryTests::wstrings", "[registry][get_registry_string]")
+            REQUIRE(e.GetErrorCode() == HRESULT_FROM_WIN32(ERROR_UNSUPPORTED_TYPE));
+        }
+    }
+    SECTION("set_value_nothrow/get_value_string_nothrow: with string key")
```
I think some of these tests are still gated on exceptions-enabled. Should we extract them?

citelao commented on include/wil/registry_basic.h:
```diff
@@ -38,1865 +36,1779 @@
-            {
-                return set_value(nullptr, value_name, value);
-            }
+                typename err_policy::result open_key(_In_opt_ PCWSTR subKey, _Out_ HKEY* hkey, ::wil::reg::key_access access = ::wil::reg::key_access::read) const
```
I don't think we should expose unsafe functions like this. We should always return managed types.

citelao commented on include/wil/registry_basic.h:
```diff
@@ -38,1865 +36,1779 @@
+        template <>
+        inline void set_value(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, const ::std::vector<::std::wstring>& data)
+        {
+            return ::wil::reg::set_value_multistring(key, subkey, value_name, data);
```
Does this mean `reg_view`'s `set_value` won't work for `set_value<std::vector<std::wstring>>`?

keith-horton replied:
right. not without reworking a bunch of stuff.

keith-horton committed:
Addressing a couple of pr comments

keith-horton committed:
Update to use std::optional for all try_ functions

keith-horton committed:
Fixed all the invalid-type get* tests

jonwis committed:
Fixes unused padding and uninitialized variable (#299)

* Ensure out-param variable is initialized

* Convert padding bytes to real

keith-horton committed:
Fixes and unit tests for BSTR and unique_bstr

keith-horton committed:
Made updates and fixes to support cotaskmem_string.
+ unit tests

keith-horton committed:
Missed some constexpr opportunities.
Verified all registry tests pass with pageheap (appverifier)

keith-horton committed:
Fixed support for byte vectors.
Added more tests and found a couple more bugs.
Added additional constexpr for callers that don't need the extra code.

keith-horton committed:
Moving functions to registry.h and renaming to following existing naming patterns.

keith-horton committed:
made significant headway to adding Doxygen comments to public functions.

Found some gaps with expand_string and started to cover them.

keith-horton committed:
More updates to comments, adding missing functions

CasualPokePlayer committed:
Use wistd::min instead of Windows.h min (#305)

The latter is not guarenteed to be available for some projects which have NOMINMAX defined somewhere before a Windows.h include.

keith-horton committed:
Fixing comments with regards to value_or() behaviors for strings + adding unit tests to verify these compile as expected.

Added cotaskmem_string missing methods.

keith-horton committed:
More documentation.
Covered a few more gaps; cleaned up some template code.

jonwis committed:
De-split tracelogging macros (#300)

* Switch to init-once helper

* General reformatting

* Merge the 'part b fields' writer macros

keith-horton committed:
Lots more documentation

mcooley committed:
Add RECT helpers (#280)

* Add RECT helpers

* Remove versions which take a pointer

* Add rect_from_size

* Oops, fix test

* Another oops

---------

Co-authored-by: Jon Wiswall <jonwis@microsoft.com>

keith-horton committed:
Almost done with documentation

talagrand committed:
Introduce make_variant_bstr* (#297)

* Introduce make_variant_bstr* to make a unique_variant that contains a BSTR in a single step

* Switched implementation slightly

---------

Co-authored-by: Jon Wiswall <jonwis@microsoft.com>
Co-authored-by: Duncan Horn <40036384+dunhor@users.noreply.github.com>

keith-horton committed:
Finished remaining functions.
Made another pass through to fix up Doxygen comments.

keith-horton committed:
Minor comment updates to show how the get_value functions returning unique_* types can also be directly assigned to shared_* types since the shared_* types have a c'tor taking a unique_* type as an R-value.

jonwis committed:
Use matched allocator/freer for token info (#306)

* Use matched allocator/freer for token info

Fixes #276

* Missed file

---------

Co-authored-by: Jon Wiswall <jonwis@ntdev.microsoft.com>

keith-horton committed:
Addressing discrepancies in naming - adding missing functions where necessary, and renaming where necessary.

m417z committed:
Add unique_evt_handle to resource.h (#311)

keith-horton committed:
Addressed feedback to cleanup Doxygen comments.
Removed SDDL usage.
Discovered gaps in string variations and filled them.

keith-horton committed:
Updated registry get* string functions. Added gaps that were found.
Updated and added corresponding unit tests.

keith-horton committed:
Minor cleanup.

dunhor committed:
Update most non-void returning const member functions to be [[nodiscard]] (#314)

* Update most non-void returning 'const' member functions to be [[nodiscard]]

* Update piplein to run tests independently and clean outputs in between to save on disk space

dunhor committed:
Fix natvis for wistd types (#315)

* Fix natvis for wistd types

* Link in the natvis by default

* Satisfy Visual Studio's expected syntax

m417z committed:
Add a throwing variant of GetSystemDirectoryW (#312)

Co-authored-by: Jon Wiswall <jonwis@microsoft.com>
Co-authored-by: Duncan Horn <40036384+dunhor@users.noreply.github.com>

citelao committed:
[Doc] Add missing info to building/testing section of README (#307)

* Add missing info to building/testing section of README

* fixing prereqs links

* refactor into sections

* describe open folder

* nit: caps

* typo

* Link to Terminal env variables bug

---------

Co-authored-by: Duncan Horn <40036384+dunhor@users.noreply.github.com>

keith-horton committed:
Cleaned up a bit mroe after getting feedback.

dunhor committed:
Work around noexcept-in-type-system VS2017 bug (#316)

dunhor committed:
Sync with the OS (#319)

* Sync with OS commit b58d62a22e8b065355e83f17d32f446bac7d5e75

* Fix build issues

* Fix test

* Merge with OS commit 2f20a1d7f273c9242cff4bb642ac0536833c0e75

* Preprocessor fix for Tracelogging.h

* EBO to make UBSan happy

* Some existing consumers can't consume win32_helpers

* Build issue; apparently [[nodiscard]] causes issues with coroutines

citelao committed:
initial work

citelao committed:
fn ptrs

citelao committed:
get, too

citelao committed:
generalizes

citelao committed:
trying to extract...

citelao committed:
array

citelao committed:
(fails) generalize to qword

citelao committed:
prototype vectors and strings

citelao committed:
hmm

citelao committed:
more hmm

citelao committed:
hmm

citelao committed:
generic test

citelao committed:
test nonexistent

citelao committed:
hmm

citelao committed:
string value

citelao committed:
remove redundant tests

citelao committed:
simpler

citelao committed:
remedy note to self

citelao committed:
test sugar methods?

citelao committed:
test sugar methods, open key

citelao committed:
remove redundant tests

citelao committed:
remove hkey from signature

citelao committed:
test sugar methods, string key

citelao committed:
remove redundant tests

citelao committed:
test throwing sugar methods, partial, open key

citelao committed:
complete sugar tests, open key

citelao committed:
remove redundant tests

citelao committed:
add missing test

citelao committed:
add template get set

citelao committed:
template get set with string key

citelao committed:
typos

citelao committed:
remove redundant tests

citelao committed:
try_get

citelao committed:
remove redundant tests

citelao committed:
forgotten

citelao committed:
playing with strings

citelao committed:
cotaskmem_string

citelao committed:
(crashes?) generic string tests

citelao committed:
remove redundant tests

citelao committed:
throwing tests

citelao committed:
remove redundant tests

citelao committed:
these don't crash

citelao committed:
test generic version

citelao committed:
remove redundant tests

citelao committed:
nits

citelao committed:
simplify error checking

citelao committed:
expanded string types

citelao committed:
more expanded string

citelao committed:
try_ expanded

citelao committed:
trying cleanup...

citelao committed:
please be shorter

citelao committed:
shorter

citelao committed:
share

citelao committed:
simpler?

citelao committed:
more refactor

citelao committed:
simplify byte vector

citelao committed:
byte vector refactored

citelao committed:
hmm!

citelao committed:
consolidate

citelao committed:
re-add optional

citelao committed:
string generic

citelao committed:
try string tests

citelao committed:
fix typo

citelao committed:
cleanup

citelao committed:
rename

citelao committed:
refactor testers

citelao committed:
hmmmmm

citelao committed:
Merge remote-tracking branch 'origin/master' into user/besto/registry_helpers_horton_v

citelao committed:
trying templated tests

citelao committed:
fix gcc compiler error

citelao committed:
fix some test typos

citelao committed:
complete end to end

citelao committed:
subkey...

citelao committed:
fix deliberate error

citelao committed:
test plain ol gets...

citelao committed:
add subkey test

citelao committed:
nothrow tests

citelao committed:
nothrow subkey

citelao committed:
multistring!

citelao committed:
test string invalid?

citelao committed:
hmm

citelao committed:
remove a lot of old tests

citelao committed:
remove more redundant tests

citelao committed:
todo

citelao committed:
add remaining generic tests

citelao committed:
remove unused vars

citelao committed:
finish simplifying multistring tests

citelao committed:
catalogue all todos

citelao committed:
Merge remote-tracking branch 'mine/user/besto/refactor_tests' into user/besto/registry_helpers_horton_v

citelao committed:
remove ignored variable to fix linting

keith-horton committed:
Fixed unit tests and filled gaps in some missing functions.

keith-horton committed:
missed a variation of try_get_value to block unique types

citelao committed:
remove unused variable

citelao committed:
gate a debug-only variable

citelao committed:
documentation

citelao committed:
replace linker error with helpful compiler error

citelao committed:
better compiler errors when using invalid string types

citelao committed:
fix unused var error (oops!)

keith-horton committed:
Addressed feedback + added a few more tests.

keith-horton committed:
Addressed feedback - cleanup and consistency updates.

keith-horton committed:
Addressed feedback.

keith-horton committed:
Made updates to reuse the caller's string buffers from shared* and unique* types when possible.

keith-horton committed:
Covered missing vector<string> functions + enabled those unit tests.

keith-horton committed:
Added test coverage as we now reuse the caller's buffer when we can.

keith-horton committed:
Clarified create vs. open, and when a subkey can be null or not.

keith-horton committed:
removing named variables for the length reg query info api.

keith-horton committed:
Addressing feedback to fix throwing/no-throwing exception specifications.

keith-horton committed:
Formatting + a few points of feedback.

keith-horton committed:
Addressed more feedback. Reworked get_value_with_type so the logic follows much simpler.

keith-horton committed:
Addressed feedback.

citelao commented:
Abandoning for https://github.com/microsoft/wil/pull/298
